File: LeetCode/1348. Tweet Counts Per Frequency/solution.py
# import bisect

# class TweetCounts:

#     def __init__(self):
#         self.times = defaultdict(set)
#         self.mp = {'minute':60, 'hour':3600, 'day':3600*24}

#     def recordTweet(self, tweetName: str, time: int) -> None:
#         self.times[tweetName].add(time)
        

#     def getTweetCountsPerFrequency(self, freq: str, tweetName: str, startTime: int, endTime: int) -> List[int]:
#         delta = self.mp[freq]
#         arr = self.times[tweetName]
#         res = []
#         for t in range(startTime,endTime+1):
#             idx = (t-startTime)//delta
#             while idx >= len(res):
#                 res.append(0)
#             if t in arr:
#                 res[-1] += 1

#         return res

class TweetCounts:

    def __init__(self):
        self.a = defaultdict(list)

    def recordTweet(self, tn: str, time: int) -> None:
        bisect.insort(self.a[tn], time)
       
    def getTweetCountsPerFrequency(self, freq: str, tn: str, startTime: int, endTime: int) -> List[int]:
        delta = 60 if freq == 'minute' else 3600 if freq == 'hour' else 86400
        i = startTime
        res = []
        while i <= endTime:
            j = min(i + delta, endTime+1)
            res.append(bisect_left(self.a[tn], j) - bisect_left(self.a[tn], i))
            i += delta
        return res
            
        


# Your TweetCounts object will be instantiated and called as such:
# obj = TweetCounts()
# obj.recordTweet(tweetName,time)
# param_2 = obj.getTweetCountsPerFrequency(freq,tweetName,startTime,endTime)

File: LeetCode/1348. Tweet Counts Per Frequency/solution.cpp
class TweetCounts {
public:
    TweetCounts() {}
    
    void recordTweet(string tweetName, int time) {
        record[tweetName][time]++;
    }
    
    vector<int> getTweetCountsPerFrequency(string freq, string tweetName, int startTime, int endTime) {
        int f = 1;
        f *= (freq == "minute") ? 60 : 1;
        f *= (freq == "hour") ? 60 * 60 : 1;
        f *= (freq == "day") ? 60 * 60 * 24 : 1;

        vector<int> ans;
        int t = startTime;
        while (t <= endTime) {
            auto bg = record[tweetName].lower_bound(t);
            auto ed = record[tweetName].upper_bound(min(t + f - 1, endTime));
            int cnt = 0;
            for (auto it = bg; it != ed; it++) {
                cnt += it->second;
            }
            ans.push_back(cnt);
            t += f;
        }
        return ans;
    }

private:
    unordered_map<string, map<int, int>> record;
};


// 作者：ikaruga
// 链接：https://leetcode-cn.com/problems/tweet-counts-per-frequency/solution/5334-by-ikaruga/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。